### PR TITLE
[WIP] [Windows] List all running processes

### PR DESCRIPTION
**Pull request by adrian-kong**
List all runnable processes following doc:
https://learn.microsoft.com/en-us/windows/win32/toolhelp/taking-a-snapshot-and-viewing-processes

### CODE REVIEW

File: src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java [hunk 1]
```diff
@@ -68,7 +60,54 @@ public static int getCurrentProcessId() {
...
+                JAVA_INT.withName("cntThreads"),
+                JAVA_INT.withName("th32ParentProcessID"),
+                JAVA_INT.withName("pcPriClassBase"),
+                JAVA_INT.withName("dwFlags"),
+                MemoryLayout.sequenceLayout(MAX_PATH, JAVA_CHAR.withName("szExeFile"))
+            );
+            // size is 560
+            var pe32 = MemorySegment.allocateNative(processEntry32, MemorySession.global());
+//            var pe32 = alloc.allocate(556);
+            var ret = (int) first.invoke(snapshot, pe32);
```
Comment (adrian-kong) [inline src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: 🤔

Comment (adrian-kong) [reply src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: have no clue why its returning error 24 - The program issued a command but the command length is incorrect.

when explicitly stating pe32.dwSize = 900 (should be more than enough) 🤔

Comment (dbwiddis) [reply src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: > When taking snapshots that include heaps and modules for a process other than the current process, the CreateToolhelp32Snapshot function can fail or return incorrect information for a variety of reasons. For example, if the loader data table in the target process is corrupted or not initialized, or if the module list changes during the function call as a result of DLLs being loaded or unloaded, the function might fail with ERROR_BAD_LENGTH or other error code. Ensure that the target process was not started in a suspended state, and try calling the function again.

Comment (adrian-kong) [reply src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: Hmmm...
Does that mean the snapshot address is incorrect or the process entry is incorrect? 
What does it mean by 'suspended' state? - How the application was ran - maven?

> and try calling the function

Just tried this inside the Exception clause - should be retry calling the function but still returns error

I'm not sure if I'm setting `pe32.dwSize` correctly 
- Get this same error on c++ by assigning smaller dwSize as well as [here (stackoverflow)](https://stackoverflow.com/a/14492379)

or the difference between `JAVA_INT` and `JAVA_FLOAT` - [DWORD](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/262627d8-3418-4627-9218-4ffe110850b2) both seem 32 bit and seems to work interchangeably

EDIT: fixed

Comment (dbwiddis) [reply src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: > or the difference between JAVA_INT and JAVA_FLOAT - [DWORD](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/262627d8-3418-4627-9218-4ffe110850b2) both seem 32 bit and seems to work interchangeably

They both represent 4 bytes and in theory could be used interchangeably, just like you could replace JAVA_DOUBLE with JAVA_LONG.  Might confuse people looking at your code. :)

Really most of the memory code is just dealing with "data structure size" and "data structure offset".  The constants are for convenience.

My pet peeve is that they took the ADDRESS size (pointer size) and documented it as size_t size.  The two are not necessarily equal!  Ah, well.

Comment (dbwiddis) [reply src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: Also, I'm not sure I ever used the snapshot for process info in OSHI due to the race condition mentioned.  I think it's useful to get a list of PIDs but there are many other (better) ways to get process info including the performance counters.

Comment (adrian-kong) [reply src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: Closing in favour of wtsapi32, which seems like how oshi did it.

(not sure if there are better ways)

PS: what would be the race condition? is it enumerating this snapshot sequentially - time consuming being outdated snapshot if doing this frequently?

Comment (dbwiddis) [reply src/main/java/ooo/oshi/foreign/windows/Kernel32Library.java]: > PS: what would be the race condition? is it enumerating this snapshot sequentially - time consuming being outdated snapshot if doing this frequently?

I'm just referring to what the docs say about modules being loaded and unloaded.  

In general for all OS's there's always race conditions involved between the initial collection of active processes and the collection of detailed data for them.

I know we went through a few iterations of various sources of Windows data, but eventually it seems that between the PDH counters (or registry shortcut) and WTS calls we get all we need, except for command lines.

**Pull request closed without merging**